### PR TITLE
Enable support for symlinks

### PR DESCRIPTION
**Pull request by jeroenpf**
Related to https://github.com/Automattic/dotcom-forge/issues/8817

## Proposed Changes

This PR aims to ensure that symlinks found inside of a site directory are working. We do this by first scanning and later watching the file system for symlinks and then creating mounts for the targets of each found symlink.

Since symlink targets inside of the NODEFS are not always predictable, we need to get the actual target of the link as seen by the FS inside the PHP runtime. When we determine the correct target, we create a mount point that points to the `realpath` of the symlink on the host system, essentially bypassing the need for chained links. (e.g.`link -> link 2 -> target` )

We also need to ensure that we keep a reference count of how many links point to a specific mounted target. It is possible that multiple symlinks point to the same target. As we are trying to unmount unnecessary targets, we need to be sure no other links point to it when we attempt to delete it.

Additionally, we need to ensure that we do not delete pre-existing files or directories at the target path. E.g. if a target already existed before we found a link pointing to it, we never want to remove it if the symlinks are removed.

**Potential issue:**
I found that deleting a symlink correctly removes files and links from the NODESF but when trying to load it via the browser, it throws an error of file not found. I think this might be related to how PHP or the underlying server is caching things or perhaps something related to the NODEFS. We might need to investigate that at some point.

## Testing Instructions

- Create a Site
- Create a file somewhere on your filesystem, place something like `<?php echo time();` in that file.
- Now, create a symlink from your Studio site directory to the file you just created by running `ln -s ~/your-file.php test.php` - this creates a symlink `test.php` that points to a file you created elsewhere.
- Try to load that file by requesting it in the browser `http://localhost:<port>/test.php` - you should see the current timestamp or the output of whatever you placed in the file.

Extra scenarios to test:
- Download a plugin somewhere and extract it, create a symlink inside `wp-content/plugins` to that plugin and see if it appears in wp-admin.
- Try adding a symlink before starting the site
- Try adding a symlink after the site has started
- Try to see if the symlink still works after 400 requests (this is when the PHP instance rotates)
- Try to delete a symlink while the site is running and verify that the file is no longer available.

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: vendor/wp-now/src/symlink-manager.ts
Comment (jeroenpf) [inline vendor/wp-now/src/symlink-manager.ts]: I realize that this is probably problematic. Will need to look into this.

File: vendor/wp-now/src/symlink-manager.ts
Comment (fluiddot) [inline vendor/wp-now/src/symlink-manager.ts]: Nitpick: Usually, type names start capitalized.
```suggestion
type MountedTarget = {
```

File: vendor/wp-now/src/symlink-manager.ts
Comment (fluiddot) [inline vendor/wp-now/src/symlink-manager.ts]: Nitpick: These two imports can be merged.

```suggestion
import fs, { FileChangeInfo } from 'node:fs/promises';
```

File: vendor/wp-now/src/symlink-manager.ts
Comment (fluiddot) [inline vendor/wp-now/src/symlink-manager.ts:1]: I noticed that the formatting of this file differs a bit from other files (e.g. spaces between brackets). I wonder if we could reformat it.

File: vendor/wp-now/src/symlink-manager.ts
Comment (fluiddot) [inline vendor/wp-now/src/symlink-manager.ts]: Nitpick: These two properties could be initialized in the constructor, as we do for the `symlink`.

File: vendor/wp-now/src/symlink-manager.ts
Comment (fluiddot) [inline vendor/wp-now/src/symlink-manager.ts]: `deleteMountedTarget` function deletes the symlink based on the reference count for the file. However, in this case, we are forcing the deletion. Shouldn't we rely on that function to remove it?

Following the above, maybe we should rename `deleteMountedTarget` to `removeSymlink`.

File: vendor/wp-now/src/symlink-manager.ts
Comment (fluiddot) [inline vendor/wp-now/src/symlink-manager.ts]: If I understand this correctly, items of `symlinks` are unique so we could remove them independently to the reference count. E.g.:

- Symlink 1:
  - `filename`: `wp-content/plugins/plugin-1.php`
  - `vfsTarget`: `/var/www/html/wp-content/plugins/hello.php`
- Symlink 2:
  - `filename`: `wp-content/plugins/plugin-2.php`
  - `vfsTarget`: `/var/www/html/wp-content/plugins/hello.php`

If I call `deleteMountedTarget('wp-content/plugins/plugin-2.php')`, that symlink can be removed safely.

File: vendor/wp-now/src/wp-now.ts
Comment (fluiddot) [inline vendor/wp-now/src/wp-now.ts]: I'm not seeing this console log when stopping a site. Looks like it's being triggered when rotating the PHP runtime. Considering that the watcher is created when starting a site, what do you think about stopping it with the site?

File: vendor/wp-now/src/symlink-manager.ts
Comment (fluiddot) [inline vendor/wp-now/src/symlink-manager.ts]: `fs.access` can fail due to other causes. I wonder if we should only return `false` in case we encounter the "file not found" error.

```suggestion
		try {
			await fs.access(filename, fs.constants.F_OK);
			return true;
		} catch (error) {
			if (error.code === 'ENOENT') {
				return false;
			}
			// Rethrow unexpected errors
			throw error;
		}
```

**Review by fluiddot — COMMENTED** [review]
Amazing work @jeroenpf 🏅 ! 

I tested different scenarios and it worked as expected. I managed to symlink a local Gutenberg repository in a site, and I was able to debug and modify blocks on live 🎊.

The only issue I found is that seems the watcher to listen for file changes is not deallocated when stopping a site. Apart from this, the changes look good to me.

File: vendor/wp-now/src/wp-now.ts
Comment (jeroenpf) [reply vendor/wp-now/src/wp-now.ts]: This may be another issue. I assumed that when we stopped the site (stop the server) it would also clean up the PHP instance (exiting it) but it does not. We should always clean up resources when they are no longer needed as not doing so may lead to memory leaks.

I think the locality of the symlink manager makes it unwieldy to hook into the stop server/stop site callback. Symlinks are closely related to the PHP instance and are recreated on rotation of the instance. I think it is better to hook into the lifecycle events of the PHP instance.

The other option to consider is to reuse the same Symlink manager instance when PHP rotates and have a way to remount the symlink targets on rotation by using the same instance of the symlink manager. We could then return the instance of the symlink manager and use in the `stopServer` method. I prefer an event-based approach though.

What do you think? Should we add a `php.exit()` call when we stop the server?

File: vendor/wp-now/src/symlink-manager.ts
Comment (jeroenpf) [reply vendor/wp-now/src/symlink-manager.ts]: Thanks, you are right. It makes more sense to delete the actual symlink entry from the map in the `deleteMountTarget` method because it always has to happen anyway.

I would not rename it to `removeSymlink` though because the symlink is not what we are really removing, we are removing the mount to which the symlink pointed. The symlink map is just a way to internally keep track of symlinks and their link targets.

File: vendor/wp-now/src/symlink-manager.ts
Comment (jeroenpf) [reply vendor/wp-now/src/symlink-manager.ts]: Thanks for pointing that out. I have moved the initialization to the constructor for consistency.

File: vendor/wp-now/src/symlink-manager.ts
Comment (jeroenpf) [reply vendor/wp-now/src/symlink-manager.ts:1]: The file is located inside `vendor/wp-now/src` and there we seem to be using a different code style. In fact, we have a `.prettierignore` file in the root and in the `wp-now` directory that explicitly tells us not to apply prettier styles so my editor is not applying them.

I think it is a bit strange we have this vendor/wp-now directory but maybe this is a common pattern in node projects? What do you suggest we do with this?

File: vendor/wp-now/src/start-server.ts
Comment (jeroenpf) [inline vendor/wp-now/src/start-server.ts:85]: I've added the `php.exit()` call here to ensure we clean up the instance and that will trigger the event handler that we added for the symlink manager as well. @fluiddot

**Comment by jeroenpf** [comment]
@fluiddot 
I've done some refactoring of the symlink manager class and moved it to the lib directory. It did not make much sense to me to keep making changes in the vendor folder.

I've also solved the issue with the watcher not being stopped properly and also return a promise from `startWatching` which makes more sense to me. Additionally, the AbortController triggers an error that we should handle separately as an 'expected' error. This is also documented in the nodejs reference documentation.

File: vendor/wp-now/src/symlink-manager.ts
Comment (fluiddot) [reply vendor/wp-now/src/symlink-manager.ts:1]: > The file is located inside `vendor/wp-now/src` and there we seem to be using a different code style. In fact, we have a `.prettierignore` file in the root and in the `wp-now` directory that explicitly tells us not to apply prettier styles so my editor is not applying them.
> 
> I think it is a bit strange we have this vendor/wp-now directory but maybe this is a common pattern in node projects? What do you suggest we do with this?

I realized after finding [this comment](https://github.com/Automattic/local-environment/pull/198#discussion_r1540910017) that it was me who disabled the styling here 😅 . The reason was that, since [`wp-now` package seems to use different formatting](https://github.com/WordPress/playground-tools/blob/trunk/.prettierrc) and we might update the files from there, we'd disable it until deciding what to do. At this point, I think we haven't made a decision so we can keep it as-is. However, I'll open a request with a proposal to use the same formatting and update all files.

File: vendor/wp-now/src/wp-now.ts
Comment (fluiddot) [reply vendor/wp-now/src/wp-now.ts]: Oh, good catch! We probably overlooked destroying the PHP instance in this logic, which as you pointed out, might be producing memory leaks.

> What do you think? Should we add a `php.exit()` call when we stop the server?

Yes, that makes total sense.  In some cases, I experienced that if the site is busy processing PHP requests, the site takes seconds to stop. This will probably improve that issue.

File: vendor/wp-now/src/symlink-manager.ts
Comment (fluiddot) [reply vendor/wp-now/src/symlink-manager.ts:1]: For reference, here's the PR to update the formatting: https://github.com/Automattic/studio/pull/547

Comment (jeroenpf) [reply vendor/wp-now/src/symlink-manager.ts:1]: Thanks! For now i actually moved the file to the lib directory so its properly styled. I think we should actually move the vendor/wp-now into the library over time because it has already diverged too much from the original wp-now and I think it would be better to not have that dependency anyway as it just means another dependency we need to maintain.

File: src/lib/symlink-manager.ts
Comment (fluiddot) [inline src/lib/symlink-manager.ts]: Nitpick: Since the filename and class already references symlink manager, we can probably remove this title.

```suggestion
```

File: src/lib/symlink-manager.ts
Comment (fluiddot) [inline src/lib/symlink-manager.ts]:
```suggestion
		// Check if the file exists
```

File: src/lib/symlink-manager.ts
Comment (fluiddot) [inline src/lib/symlink-manager.ts]:
```suggestion
		// If the file does not exist, it was deleted. Check if it was a symlink and remove it.
```

File: vendor/wp-now/src/symlink-manager.ts
Comment (fluiddot) [reply vendor/wp-now/src/symlink-manager.ts]: > I would not rename it to `removeSymlink` though because the symlink is not what we are really removing, we are removing the mount to which the symlink pointed. The symlink map is just a way to internally keep track of symlinks and their link targets.

But the mountpoint is only removed if there are no other symlinks referencing it, right? My rationale for renaming it is that we have the function `addSymlink`. When a new symlink is created in the file system, we call that function. So, when it's removed, I'd expect that we call `removeSymlink`. Besides, removing a mount is a side effect of removing symlinks.

File: src/lib/symlink-manager.ts
Comment (fluiddot) [inline src/lib/symlink-manager.ts]: In other areas of the code, we check undefined values with the following approach:

```suggestion
		if ( typeof mountInfo === 'undefined' ) {
```

File: src/lib/symlink-manager.ts
Comment (fluiddot) [inline src/lib/symlink-manager.ts:199]: In other areas of the code, we check undefined values with the following approach:

```suggestion
		if ( typeof vfsTarget === 'undefined' ) {
```

**Review by fluiddot — COMMENTED** [review]
On macOS the PR is working great, however, symlinks are not working on Windows. Here's the results I'm getting:

### How to create symlink
1. Open Powershell session with admin privileges.
2. Go to `<SITE_FOLDER>/wp-content/plugins`.
3. Run the command `New-Item -Path new-plugin.php -ItemType SymbolicLink -Value hello.php`.

### Results

- When adding a symlink, [this statement](https://github.com/Automattic/studio/pull/510/files#diff-c7a9483ceff52ddf6338248a4a2c9141a54e18edd734dccd8f00c578cdfeb1dcR149) always return `false`. I investigated further and this is a problem with the path separator used. When using `path.join`, the package picks up the one based on the OS, in this case, Windows. However, the PHP environment is Unix-based, so Windows paths won't work.
- I forced using Unix separate when generating `vfsPath`, but the path returned when reading the symlink ([ref](https://github.com/Automattic/studio/pull/510/files#diff-c7a9483ceff52ddf6338248a4a2c9141a54e18edd734dccd8f00c578cdfeb1dcR154)) is returning a wrong value. E.g. `vfsTarget = '/var/www/html/wp-content/plugins/C:\\Users\\Carlos\\Sstudio\\my-site\\wp-content\\plugins\\hello.php'`.

@jeroenpf I'm open to tackle this in a separate PR, in case we'd like to only provide support for symlinks in macOS. WDYT?

**Comment by jeroenpf** [comment]
> On macOS the PR is working great, however, symlinks are not working on Windows. Here's the results I'm getting:
> 
> ### How to create symlink
> 1. Open Powershell session with admin privileges.
> 2. Go to `<SITE_FOLDER>/wp-content/plugins`.
> 3. Run the command `New-Item -Path new-plugin.php -ItemType SymbolicLink -Value hello.php`.
> 
> ### Results
> * When adding a symlink, [this statement](https://github.com/Automattic/studio/pull/510/files#diff-c7a9483ceff52ddf6338248a4a2c9141a54e18edd734dccd8f00c578cdfeb1dcR149) always return `false`. I investigated further and this is a problem with the path separator used. When using `path.join`, the package picks up the one based on the OS, in this case, Windows. However, the PHP environment is Unix-based, so Windows paths won't work.
> * I forced using Unix separate when generating `vfsPath`, but the path returned when reading the symlink ([ref](https://github.com/Automattic/studio/pull/510/files#diff-c7a9483ceff52ddf6338248a4a2c9141a54e18edd734dccd8f00c578cdfeb1dcR154)) is returning a wrong value. E.g. `vfsTarget = '/var/www/html/wp-content/plugins/C:\\Users\\Carlos\\Sstudio\\my-site\\wp-content\\plugins\\hello.php'`.
> 
> @jeroenpf I'm open to tackle this in a separate PR, in case we'd like to only provide support for symlinks in macOS. WDYT?

Good find! I will address it in this PR. We should be able to use `path.posix.join` to force posix compliant paths when dealing with paths inside WASM. I will try this out.

File: src/lib/symlink-manager.ts
Comment (fluiddot) [inline src/lib/symlink-manager.ts]: I've just realized that we already have this function defined in `lib/fs-utils.ts`:

https://github.com/Automattic/studio/blob/8e6816e2ba015c7013e681d63f9c359369df01eb/src/lib/fs-utils.ts#L7-L17

Comment (jeroenpf) [reply src/lib/symlink-manager.ts]: Thanks! I wasn't aware but I removed it now and used the existing function.

File: vendor/wp-now/src/wp-now.ts
Comment (fluiddot) [inline vendor/wp-now/src/wp-now.ts]: In other areas of the app, we check if it's running on the Windows platform with the following code:

```suggestion
	if ( process.platform !== 'win32' ) {
```

File: vendor/wp-now/src/wp-now.ts
Comment (fluiddot) [inline vendor/wp-now/src/wp-now.ts]: I noticed this import is not being used in the code, so we could remove it.

File: vendor/wp-now/src/wp-now.ts
Comment (jeroenpf) [reply vendor/wp-now/src/wp-now.ts]: I saw that and it made me wonder if at some stage it could potentially be win64?  For now I'll change it to keep it in line with the rest.

**Comment by jeroenpf** [comment]
I've added some tests and fixed some more small items based on feedback.

File: vendor/wp-now/src/wp-now.ts
Comment (fluiddot) [reply vendor/wp-now/src/wp-now.ts]: Yeah, good point. The `platform` value is provided by Node.js and there's no `win64` value ([reference](https://nodejs.org/api/process.html#processplatform)). I think even when using Windows 64 bits it references `win32`.

**Comment by fluiddot** [comment]
@jeroenpf not a blocker but it would be great if we could solve the file conflicts with `trunk`.

File: src/lib/tests/symlink-manage.test.ts
Comment (fluiddot) [inline src/lib/tests/symlink-manage.test.ts]: I think we don't need to define these properties as they will be incorporated in line 47:
```
...jest.requireActual( 'path' ),
```

File: src/lib/tests/symlink-manage.test.ts
Comment (fluiddot) [inline src/lib/tests/symlink-manage.test.ts]: I wonder if it's necessary to cast `mockWatcher` as `never`. WDYT?

File: src/lib/tests/symlink-manage.test.ts
Comment (fluiddot) [inline src/lib/tests/symlink-manage.test.ts]: I wonder if we could change the order here:
```suggestion
			await symlinkManager.startWatching();
			expect( fs.watch ).toHaveBeenCalledWith( mockProjectPath, expect.any( Object ) );
			await symlinkManager.stopWatching();
```

File: src/lib/tests/symlink-manage.test.ts
Comment (fluiddot) [inline src/lib/tests/symlink-manage.test.ts:95]: We could also check mounted targets:
```
expect(
	( symlinkManager as unknown as SymlinkManagerPrivateProperties ).mountedTargets.size
).toBe( 2 );
```

File: vendor/wp-now/src/wp-now.ts
Comment (wojtekn) [reply vendor/wp-now/src/wp-now.ts]: Win32 is the name of the platform API and not the processor architecture.

File: src/lib/tests/symlink-manage.test.ts
Comment (fluiddot) [inline src/lib/tests/symlink-manage.test.ts]: The React testing library provides a way to deal with asynchronous code via `waitFor`. Here's a potential update on on the code:

```suggestion
			await symlinkManager.startWatching();
			await waitFor( () => {
				expect( mockPHP.mkdir ).toHaveBeenCalled();
				expect( mockPHP.mount ).toHaveBeenCalled();
				const privateProps = symlinkManager as unknown as SymlinkManagerPrivateProperties;
				expect( privateProps.symlinks.has( 'newSymlink' ) ).toBe( true );
			} );
			await symlinkManager.stopWatching();
```

File: src/lib/tests/symlink-manage.test.ts
Comment (fluiddot) [inline src/lib/tests/symlink-manage.test.ts]: Nitpick:

```suggestion
			const unmountFunction = jest.fn();
			privateProps.mountedTargets.set( '/vfs/path/to/target', {
				referenceCount: 1,
				unmountFunction,
			} );
```

File: src/lib/tests/symlink-manage.test.ts
Comment (fluiddot) [inline src/lib/tests/symlink-manage.test.ts]: Similar to my comment above:

```suggestion
			await symlinkManager.startWatching();
			await waitFor( () => {
				expect( privateProps.symlinks.has( 'existingSymlink' ) ).toBe( false );
				expect( unmountfn ).toHaveBeenCalled();
			} );
			await symlinkManager.stopWatching();
```

File: vendor/wp-now/src/wp-now.ts
Comment (jeroenpf) [reply vendor/wp-now/src/wp-now.ts]: Thats good to know, thanks for the info!

File: src/lib/tests/symlink-manage.test.ts
Comment (fluiddot) [inline src/lib/tests/symlink-manage.test.ts:1]: It would be also interesting to cover the following cases:
- A mounted target is not removed if it's referenced by other symlinks.
- When mounting a target, it doesn't call `php.mkdir` if the symlink points to a preexisting path (i.e. one that is already mounted in PHP).

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 ! Awesome work @jeroenpf 🏅 !

I've tested the functionality on different platforms:
- macOS ✅ 
- Windows ✅ (Symlinks are disabled)
- Linux ✅ 

I've added some comments in the unit tests, but none should block merging the PR.

File: src/lib/tests/symlink-manage.test.ts
Comment (jeroenpf) [reply src/lib/tests/symlink-manage.test.ts:95]: Added

File: src/lib/tests/symlink-manage.test.ts
Comment (jeroenpf) [reply src/lib/tests/symlink-manage.test.ts]: Good point, it was a leftover from some things i was trying but its indeed not necessary.

File: src/lib/tests/symlink-manage.test.ts
Comment (fluiddot) [reply src/lib/tests/symlink-manage.test.ts]: Forgot to mention, but we need to import `waitFor`:
```
import { waitFor } from '@testing-library/react';
```

File: src/lib/tests/symlink-manage.test.ts
Comment (jeroenpf) [reply src/lib/tests/symlink-manage.test.ts:1]: I've added the tests :-) I'll go ahead and merge this PR once the tests pass.

**Pull request merged**
